###### Andrés Genta

In [34]:
import numpy as np 

In [35]:
print(np.version.version) 
"""
1.15.2
"""

1.24.3


'\n1.15.2\n'

In [36]:
a=np.random.randint(10, size=(2,3,5))

### Formas de generar random arrays:
1) random.randint()
2) random.random()
3) random.rand()   

In [37]:
print(a)

[[[1 3 1 5 4]
  [7 4 1 0 3]
  [1 1 3 8 1]]

 [[3 6 4 7 1]
  [8 0 1 6 1]
  [9 3 4 7 9]]]


In [38]:
b=np.ones((5,2,3))

In [39]:
print(b)

[[[1. 1. 1.]
  [1. 1. 1.]]

 [[1. 1. 1.]
  [1. 1. 1.]]

 [[1. 1. 1.]
  [1. 1. 1.]]

 [[1. 1. 1.]
  [1. 1. 1.]]

 [[1. 1. 1.]
  [1. 1. 1.]]]


In [40]:
if a.size==b.size:
    print(True)

True


### 8)

No es posible sumar "a" con "b" ya que tienen diferentes shapes.

In [41]:
c=np.transpose(b,(1,2,0))  
c.shape

(2, 3, 5)

Con el vector (1,2,0) indico que la transposición sea de la forma (a,b,c) con a<b y b<c.

In [42]:
d=a+c
print(d)
print(type(d))
d.shape

[[[ 2.  4.  2.  6.  5.]
  [ 8.  5.  2.  1.  4.]
  [ 2.  2.  4.  9.  2.]]

 [[ 4.  7.  5.  8.  2.]
  [ 9.  1.  2.  7.  2.]
  [10.  4.  5.  8. 10.]]]
<class 'numpy.ndarray'>


(2, 3, 5)

Fue posible que se realice la suma porque luego de la transposición, ambos arrays quedaron con el mismo shape.

In [43]:
print(a)
print(d)

[[[1 3 1 5 4]
  [7 4 1 0 3]
  [1 1 3 8 1]]

 [[3 6 4 7 1]
  [8 0 1 6 1]
  [9 3 4 7 9]]]
[[[ 2.  4.  2.  6.  5.]
  [ 8.  5.  2.  1.  4.]
  [ 2.  2.  4.  9.  2.]]

 [[ 4.  7.  5.  8.  2.]
  [ 9.  1.  2.  7.  2.]
  [10.  4.  5.  8. 10.]]]


Cada elemento del array "d" es superior a su correspondiente elemento en "a" por un valor de 1, dado que a "a" se le sumó el array de ones.

In [44]:
e=a*c
e

array([[[1., 3., 1., 5., 4.],
        [7., 4., 1., 0., 3.],
        [1., 1., 3., 8., 1.]],

       [[3., 6., 4., 7., 1.],
        [8., 0., 1., 6., 1.],
        [9., 3., 4., 7., 9.]]])

In [45]:
equal_result = (a == e)       #Son iguales pues el resultado solo está compuesto con elementos=True
print("Equality Comparison Result:")
print(equal_result)

Equality Comparison Result:
[[[ True  True  True  True  True]
  [ True  True  True  True  True]
  [ True  True  True  True  True]]

 [[ True  True  True  True  True]
  [ True  True  True  True  True]
  [ True  True  True  True  True]]]


In [46]:
d_max=np.max(d)
d_max

10.0

In [47]:
d_min=np.min(d)
d_min

1.0

In [48]:
d_mean=np.mean(d)
d_mean

4.733333333333333

### 15)

In [49]:
f=np.empty([2,3,5])   #No entiendo, porque no me da una matriz vacía
print(f)

[[[1. 3. 1. 5. 4.]
  [7. 4. 1. 0. 3.]
  [1. 1. 3. 8. 1.]]

 [[3. 6. 4. 7. 1.]
  [8. 0. 1. 6. 1.]
  [9. 3. 4. 7. 9.]]]


### 16)

In [50]:
f[(d > d_min) & (d < d_mean)] = 25
f[(d > d_mean) & (d < d_max)] = 75
f[d == d_mean] = 50
f[d == d_min] = 0
f[d == d_max] = 100


In [51]:
f

array([[[ 25.,  25.,  25.,  75.,  75.],
        [ 75.,  75.,  25.,   0.,  25.],
        [ 25.,  25.,  25.,  75.,  25.]],

       [[ 25.,  75.,  75.,  75.,  25.],
        [ 75.,   0.,  25.,  75.,  25.],
        [100.,  25.,  75.,  75., 100.]]])

### 17)

In [52]:
print(d)
print(f)

[[[ 2.  4.  2.  6.  5.]
  [ 8.  5.  2.  1.  4.]
  [ 2.  2.  4.  9.  2.]]

 [[ 4.  7.  5.  8.  2.]
  [ 9.  1.  2.  7.  2.]
  [10.  4.  5.  8. 10.]]]
[[[ 25.  25.  25.  75.  75.]
  [ 75.  75.  25.   0.  25.]
  [ 25.  25.  25.  75.  25.]]

 [[ 25.  75.  75.  75.  25.]
  [ 75.   0.  25.  75.  25.]
  [100.  25.  75.  75. 100.]]]


##### Obtuve lo esperado. 
##### Obs: No obtuve ningún elemento en "f" de valor 50 porque el mean de "d" es ~ 4.73 y los elementos de "d"         están en enteros.

### 18) **BONUS**

In [53]:
f_en_lista=f.tolist()
print(f_en_lista)
type(f_en_lista)


[[[25.0, 25.0, 25.0, 75.0, 75.0], [75.0, 75.0, 25.0, 0.0, 25.0], [25.0, 25.0, 25.0, 75.0, 25.0]], [[25.0, 75.0, 75.0, 75.0, 25.0], [75.0, 0.0, 25.0, 75.0, 25.0], [100.0, 25.0, 75.0, 75.0, 100.0]]]


list

In [54]:
f_nuevo=[]
for matriz in f_en_lista:
    for lista in matriz:
        for elem in lista:
            f_nuevo.append(elem)
print(f_nuevo)

for i in range(len(f_nuevo)):
    if f_nuevo[i]==0:
        f_nuevo[i]="A"
    elif f_nuevo[i]==25.0:
        f_nuevo[i]="B"
    elif f_nuevo[i]==50:
        f_nuevo[i]="C"
    elif f_nuevo[i]==75:
        f_nuevo[i]="D"
    else:
        f_nuevo[i]="E"
    
print(f_nuevo)
            

[25.0, 25.0, 25.0, 75.0, 75.0, 75.0, 75.0, 25.0, 0.0, 25.0, 25.0, 25.0, 25.0, 75.0, 25.0, 25.0, 75.0, 75.0, 75.0, 25.0, 75.0, 0.0, 25.0, 75.0, 25.0, 100.0, 25.0, 75.0, 75.0, 100.0]
['B', 'B', 'B', 'D', 'D', 'D', 'D', 'B', 'A', 'B', 'B', 'B', 'B', 'D', 'B', 'B', 'D', 'D', 'D', 'B', 'D', 'A', 'B', 'D', 'B', 'E', 'B', 'D', 'D', 'E']


In [55]:
f_final=np.resize(f_nuevo,(2,3,5))
print(f_final)
f_final.shape

[[['B' 'B' 'B' 'D' 'D']
  ['D' 'D' 'B' 'A' 'B']
  ['B' 'B' 'B' 'D' 'B']]

 [['B' 'D' 'D' 'D' 'B']
  ['D' 'A' 'B' 'D' 'B']
  ['E' 'B' 'D' 'D' 'E']]]


(2, 3, 5)

### Lo primero que hice fue pasar "f" a lista porque numpy es para trabajar con números. Luego de tenerlo en type lista, en realidad quedó como una lista con listas, por lo que hice 3 fors combinados para pasarlo a una única lista (me queda más cómodo hacerlo así que con las list comprehensions). Después cambié los números por las letras correspondientes y lo transformé en un array haciéndole un resize para que quede con el shape original de f.